### DISCUSSION TITLE
Discussion about the Testing API

### DISCUSSION THREAD

**Discussion by zero-plusplus** [zero-plusplus/vscode-autohotkey-debug#127] [Ideas]
[Testing API](https://code.visualstudio.com/api/extension-guides/testing) is now complete with [vscode 1.59](https://code.visualstudio.com/updates/v1_59#_testing-apis).

Technically, it is not clear at this time whether this API can be applied to AutoHotkey.

However, I'm very interested in testing AutoHotKey, so I've opened this discussion to get your feedback.

If it is technically feasible, are there any features you would like to see?
Also, would you be interested in a unit test library that works with v1 and v2?

---

The [experimental testing framework](https://github.com/zero-plusplus/experimental-autohotkey-testframework/tree/develop) is now available. Currently, only the minimum functionality is implemented.

It is used in this [project](https://github.com/zero-plusplus/bee.ahk/tree/alpha/test).

**Comment by SAbboushi** [comment]
I am not knowledgable about unit testing; I looked at the Testing API link you provided, but am still very unclear on what it means and how it  could provide value to AHK users.

I often wonder why you do not post on the AHK forum?  I think it would be great if we can get some of the advanced AHK users to provide input on such matters.

In the past, I think you expressed concern that your English was an obstacle.  In case that is still a concern, I would like to tell you that it has been very rare that I have had any difficulty in clearly understanding your posts i.e. your English/the DeepL translator has been very effective.

**Reply by zero-plusplus** [reply]
There are many reasons why I intentionally do not use the AHK forum, but broadly speaking, they are as follows.

1. I'm a slow writer by nature, plus I spend 30 minutes to 2 hours coordinating translations using the DeepL feature. Therefore, if I speak in a place with a large community, my days off will be gone

2. I would like to keep discussions about this extension within Github as much as possible. This is because it will be easier to search for it later. Another reason is that it is easier to receive notifications on my smartphone

3. I want to hear from people who are using this extension

---

Unit test is a way to test the input and output of a function. For example, SubStr can be tested as follows.

```ahk
#Requires AutoHotkey v1.1.33+

TestSubStr()
TestSubStr() {
  Assert(SubStr("abc", 2) == "bc") ; PASS
  Assert(SubStr("abc", 0) == "c") ; PASS
  Assert(SubStr("abc", 2) == "b") ; FAIL
}

Assert(assertion) {
  if (!assertion) {
    OutputDebug, FAIL`n
    return
  }
  OutputDebug, PASS`n
}
```

---

The Testing API is a feature to run and debug the defined tests from the GUI. 
In the above example, testSubstr can be run and debugged from the GUI.

Since a parser is needed to detect testSubStr, implementation is quite a ways off.
This is another reason why I don't post in the AHK forum.

**Comment by SAbboushi** [comment]
Thanks for the education and sharing with me your reasons.

**Reply by zero-plusplus** [reply]
If you have any questions in the future, please feel free to ask. This can lead to new ideas.
